packages/domainproperties/src/components/assay/AssayDesignerPanels.tsx

```diff
@@ -94,40 +105,74 @@ export class AssayDesignerPanels extends React.PureComponent<Props, State> {
         }
     };
 
+    onTogglePanel = (index: number, collapsed: boolean, callback: () => any) => {
+        const { visitedPanels, currentPanelIndex } = this.state;
+
+        if (!collapsed) {
+            if (!visitedPanels.contains(index)) {
+                this.setState(() => ({currentPanelIndex: index,
+                    visitedPanels: visitedPanels.push(index),
+                    firstState: false}), callback());
+            }
+            else {
+                this.setState(() => ({currentPanelIndex: index, firstState: false}), callback());
+            }
+        }
+        else {
+            if (collapsed && currentPanelIndex === index) {
+                this.setState(() => ({currentPanelIndex: undefined, firstState: false}), callback());
+            }
+            else {
+                callback();
+            }
+        }
+    };
+
     onFinish = () => {
         const { protocolModel } = this.state;
         const { beforeFinish } = this.props;
 
-        this.setSubmitting(true);
-        if (beforeFinish) {
-            beforeFinish(protocolModel);
-        }
+        this.setState((state) => ({validatePanel: state.currentPanelIndex}), () => {
+            this.setState((state) => ({validatePanel: undefined}), () => {
```

Comment:
This seems odd to have two calls to this.setState() nested directly within one another. If I am reading this correctly, the validatePanel gets set to state.currentPanelIndex and then after that state change gets set immediately to undefined. Any reason to not just go directly to undefined?